### ISSUE TITLE
Network Spawned character does not appear on client

### ISSUE THREAD

**Issue by astromediaonly** [insthync/LiteNetLibManager#7]
Im trying to learn how this basic stuff works but i need one small push in right direction.
I have server and client connected to it, i do ``Assets.NetworkSpawn(spawnObj, 0, connId);`` on server but that object does not appear on client ... i suppose i have missed some step and this part does not seems to be in docs. Im trying to spawn player character for connected client. 
Thx for any help/info.

TL;DR: I need some "checklist" what to set where / do through code to make networkspawned object appear as player on owning client

**Comment by insthync** [comment]
Did you add the prefab to `LiteNetLibAssets` → `SpawnablePrefabs` yet ?
![Screenshot_187](https://user-images.githubusercontent.com/2697550/69475259-80fa5d80-0dfd-11ea-8a11-c828e7986113.png)

**Comment by astromediaonly** [comment]
Yes, both on client and server object.

**Comment by astromediaonly** [comment]
Anything else i could have missed? (it will be probably something stupid at the end so i apologize in advance :) )

**Comment by insthync** [comment]
So...
You already set player character to `Spawnable Prefabs` then
You call `Assets.NetworkSpawn(spawnObj, 0, connId);` at server
I don't know when did you call `Assets.NetworkSpawn(spawnObj, 0, connId);`
Did you override `LiteNetLibGameManager` to changes some functionality or not?

**Comment by insthync** [comment]
For the default processes, **client** will send `EnterGame` message to **client** after connected IF `doNotEnterGameOnConnect` is `FALSE` 

Then **server** which receives `EnterGame` from **client** will send `EnterGame` message with `ConnectionId` and `ServerSceneChange` message with `OnlineScene`  to client.

Then **client** will load scene if current scene is not the same as **server**

After scene loaded **client** will send `ClientReady` message to **server**. 
This is when you can add some extra data to **server**. You can do it by extends `LiteNetLibGameManager` class and override `SerializeClientReadyExtra` function to write an extra data that sending to **server**
For Example
```
public override void SerializeClientReadyExtra(NetDataWriter writer) {
  writer.Put(characterName);
  writer.Put(characterType);
}
```
And you have to override `DeserializeClientReadyExtra` to read an extra data that receiving from **client**
```
public override void DeserializeClientReadyExtra(LiteNetLibIdentity playerIdentity, long connectionId, NetDataReader reader) {
  var characterName = reader.GetString();
  var characterType = reader.GetInt();
}
```
Then you may spawn player character in `DeserializeClientReadyExtra` function
```
public override void DeserializeClientReadyExtra(LiteNetLibIdentity playerIdentity, long connectionId, NetDataReader reader) {
  var characterName = reader.GetString();
  var characterType = reader.GetInt();
  if (characterType == 0)
  {
    Assets.NetworkSpawn(Instantiate(prefabA.gameObject, Vector3.zero, Quaternion.identity), 0, connectionId);
  } else {
    Assets.NetworkSpawn(Instantiate(prefabB.gameObject, Vector3.zero, Quaternion.identity), 0, connectionId);
  }
}
```

**Comment by astromediaonly** [comment]
Thx for response.
Yes i do, im doing it inside ``RegisterClientMessage`` callback on my ``MapServer`` version of ``LiteNetLibGameManager``. Relevant part of code follows:

```csharp
public class MapServer : LiteNetLibGameManager
{
	protected override void RegisterClientMessages()
	{
		base.RegisterClientMessages();
		RegisterClientMessage(PacketCodes.LOGIN_PLAYER_CONNECTED, (messageHandler) => {
			
			//...
			
			Guid loginKey = Guid.Parse(packet.LoginKey);
			
			//....

			GameObject spawnObj = Instantiate(GameManager.Instance.PrefabPlayerCharacter.gameObject, new Vector3(ch.Data.PosX, ch.Data.PosY, ch.Data.PosZ), Quaternion.identity);

			ServerConsole.Instance.Log(ConsoleLogType.DEBUG_INFO, "Player spawned with model " + model.gameObject.name); //this triggers

			Assets.NetworkSpawn(spawnObj, 0, OnlineConnections[loginKey]);

			// Setup subscribers
			LiteNetLibPlayer player = GetPlayer(OnlineConnections[loginKey]);
			foreach (LiteNetLibIdentity spawnedObject in Assets.GetSpawnedObjects())
			{
				if (spawnedObject.ConnectionId == player.ConnectionId)
					continue;

				if (spawnedObject.ShouldAddSubscriber(player))
					spawnedObject.AddSubscriber(player);
			}
			
			//....
		});
	}
}
```

I have copyed that spawn part from your MMO Kit (i still have to figure out how helpfull is that subscribers part but it should not be that important right now i suppose - i had included it anyway to be sure). Its not final approach, im still figuring out what is best way to structure it for me but i suppose it should be ok.
I also hope its ok i have 3 scripts derived from ``LiteNetLibGameManager`` in scene and in this case im comunicating from client one to map server one (only one is active at time for specific executable tho and both of them have that player character prefab filled into Spawnable prefabs as i mentioned before).
Sorry for not providing this info from start but im not sure what is relevant and what not so much and i wanted to spare you too much clutter im usually able to write in failed attempt to describe what am i doing :)

**Comment by astromediaonly** [comment]
Thx for that additional info, i have probably done too much stuff on my own from what i see (managing scenes specificaly) and i suppose it bypassed some necessary steps for it to work properly or something like that ... i will try to process this info and let you know if it helped. 

What i did is - client connects to my login server, login info is sent, server will responds if it was or not succesfull, if it was character is selected and login server is informed about that and if everything is ok client is disconnected from login server and connected to appropriate map server, then there is some "chat" between map server and login server to get all info about that player and if everything went correctly that spawning script i posted is executed and then i expected player object to appear on client as well which was probably oversimplification it seems. I did not messed up with any ``EnterGame`` or ``ServerSceneChange`` logic and im doing all scene change logic my self so it was not necessary for that Library to handle it for me.

**Comment by astromediaonly** [comment]
Thanks so much - pointing me to relevant code part or telling me what in general should be happening is usually enough for me to catch up :) - (im not sure if its only one but) i was able to find the issue ... its inside this:
```csharp
protected virtual void HandleClientEnterGame(LiteNetLibMessageHandler messageHandler)
{
	SendServerEnterGame(messageHandler.connectionId);
	SendServerSceneChange(messageHandler.connectionId, ServerSceneName);
	// If it is host (both client and server) it will send ready state to spawn player
	if (IsClient && (string.IsNullOrEmpty(serverSceneName) || serverSceneName.Equals(SceneManager.GetActiveScene().name)))
		SendClientReady();
}
```
it expects ``string.IsNullOrEmpty(serverSceneName)`` is the name of map client is on but its not in my case it seems (server has there stored my init map name) ... i did not used ``LoadSceneRoutine`` (which is the problem i suppose) but just done it my way using regular ``SceneManager.LoadScene`` . I will take a look if i can use it but there seems to be lot of logic and im afraid it will not work well with rest of my code. Maybe your approach is better and i should use your hooks and modify code to work with that ... either way if you have some tips on what to do and what not with it it will be appretiated :) 

*dramatic pause* - so i have done some more investigating and if i understand correctly after looking into this code:
```csharp
public override void OnStartServer()
{
	base.OnStartServer();
	// Reset client connection id, will be received from server later
	ClientConnectionId = -1;
	if (!Assets.onlineScene.IsSet() || Assets.onlineScene.SceneName.Equals(SceneManager.GetActiveScene().name))
	{
		serverSceneName = SceneManager.GetActiveScene().name;
		Assets.Initialize();
		Assets.SpawnSceneObjects();
		OnServerOnlineSceneLoaded();
	}
	else
	{
		serverSceneName = Assets.onlineScene.SceneName;
		StartCoroutine(LoadSceneRoutine(Assets.onlineScene.SceneName, true));
	}
}
```
most proper way for me would be to set ``Assets.onlineScene.SceneName`` before starting map server manager, let it do all its "magic" and put all my "server done scene loading" logic inside ``Assets.onLoadSceneFinish`` event instead of ``SceneManager.sceneLoaded`` where i have it now - am i correct?

**Comment by astromediaonly** [comment]
Ok, progress - i have rewritten it as i was mentioning and it seems its working after some fixes (dumb me forgot to add ``base.RegisterClientMessages()`` after overriding it on my client instance so it was ignoring all "system" messages from server  :/ ) ... mostly ... it even calls ``NetworkSpawn`` on client and ``OnSetOwnerClient`` on my entity but im not sure why/how cause i dont see my player object spawned on client in hierarchy. Is it something destroying it afterwards or any other idea why its not in hierarchy? 

I have tryed adding ``OnDestroy`` with debug log to it and it indeed seems its destroyed as soon as its created ... do you have any idea what can cause that? Thx for any additional help/info.

**Comment by astromediaonly** [comment]
Just to add additional info - from what i was able to debug the object is destroyed by ``subscriber.RemoveSubscribing(this, true);`` inside ``RebuildSubscribers`` in script ``LiteNetLibIdentity``. Can i get some info how that subscribers logic should work? I have probably missed some setup or something. :/ 

Sorry for having so many questions, it has steep learning curve for somebody new to it but i feel im making progress.

**Comment by astromediaonly** [comment]
Uh, i was finally able to figure out what was happening - i had prepaired visibility checker with some default settings on my identity (because i was setting it 1:1 what was in MMO Kit when i was not able to spawn my player during previous issue) and it resulted in not adding player to ``newSubscribers`` array during initial rebuild when ``OnRebuildSubscribers`` was called ... i think. When i removed it it stays spawned it seems. But if you can point me to some information on how those subscribers and visibility checker is supposed to work it would be still nice - i have feeling i will need it too eventually.

**Comment by astromediaonly** [comment]
Hi, sorry for writing again but any info about expanding the docs? :) (some "checklist" examples of commonly done stuff like spawning different objects would be nice :) ) - sadly im fighting with spawning again - now for npcs ... maybe you can see what i missed this time - i had done this:

1. Added on client and server entity to spawnable prefabs
2. When ``OnSceneLoaded`` is called after server is started i do ``Assets.NetworkSpawn(entity.gameObject);`` (i had omited those two params used for player character - i suppose server controlled object does not need owner connectionId

But for some reason i dont see my spawned entity on client and it does not seems to be because any of problems i was solving with player version - is there some difference for nonplayer driven entity? Do i need visibility checker in this case or i should see everything networkspawned by default?
Thx in advance for any hint.

**Comment by astromediaonly** [comment]
Ok, after sleep i have made additional test and problem seems to be execution order-ish: when i use that ``OnSceneLoaded`` it does not work but if i give there ``IEnumerator`` and start it as coroutine and there wait for 10s it works it seems so my question is - when is safe to spawn objects after server starts? :)

**Comment by insthync** [comment]
Spawn object in function `OnServerOnlineSceneLoaded()` must be works, if it's not work it might be a bug, I will check it.

**Comment by astromediaonly** [comment]
Its technicaly handler attached to ``Assets.onLoadSceneFinish`` ... sorry if i mistaken it for something else. 
I do this:
```
private void OnEnable()
{
	Assets.onLoadSceneFinish.AddListener(OnSceneLoaded);
}

public void StartMapServer()
{
	Assets.onlineScene.SceneName = Map.Scene.SceneName;
	StartServer();
}

void OnSceneLoaded(string sceneName, bool isOnline, float a)
{
	if (sceneName.Equals(Map.Scene.SceneName)) 
	{
		//this calls method in another file containing what is between next comments
		this.InvokeInstanceDevExtMethods("OnMapLoaded"); 
		
		///
		//... some code here to setup that entity
		Assets.NetworkSpawn(entity.gameObject);
		///
	}
}
```

In case its problem with execution order or something ... it gets called fine but does nothing so i guess there is atleast one bug there - if server is not ready yet, it should probably throw some exception i suppose? :)

Hope this additional info helps. Thanks for your time.

**Comment by insthync** [comment]
Ahh, I see, I think I should change the function call order, this was calling before initialize prefabs and scene objects.

**Comment by astromediaonly** [comment]
Nice to hear its not my fault this time :) ... let me know after you will (find a way how to) fix it ... it does not hurry tho, my 10s wait hack works for now so i can continue with my project meanwhile. Thx for info.

**Comment by insthync** [comment]
I've made the update, you can pull from git repository.

**Comment by astromediaonly** [comment]
That was quick :) ... it seems to be working, thanks again.
